docs/proposals/prop-apb-gen-creds.md

```diff
@@ -0,0 +1,41 @@
+# APB Generate Credentials Secret Proposal
+
+## Introduction
+
+This proposal aims to improve the broker process by which we extract
+credentials from bindable [Ansible Playbook
+Bundles (APB)](https://github.com/ansibleplaybookbundle/ansible-playbook-bundle)
+by allowing the APB to generate a secret with needed credentials instead of
+using `kubectl exec` into the running pod to grab credentials from a file.
+
+Reference Issue #544
+
+## Problem Description
+
+The broker should seek to use available kubernetes API calls to get necessary
+information whenever possible instead of relying on `kubectl exec`. Now that
+the broker supplies pod name and namespace information to running APBs via the
+[Kubernetes Downward
+API](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/),
+APBs have all they need to generate secrets inside the APB sandbox namespace to
+be retrieved later by the broker.
+
+## Implementation Details
+
+- [ ] Update the [`asb_encode_binding`
+module](https://github.com/ansibleplaybookbundle/ansible-asb-modules/blob/master/library/asb_encode_binding.py)
+to create a kubernetes secret in the APB sandbox namespace.
```

Comment:
Are there any error conditions we need to be able to handle? Anything that could cause the information not to appear in the secret?

Reply:
Could you elaborate more on this. Off the top of my head the only scenario that I would be concerned about would be if we fail to create the secret. I'm struggling to think of a scenario where we would be able to create the secret but it would be empty.

Reply:
I don't know :) just trying to come up with worst case scenarios. Can creating a secret fail? if we get errors what will happen? will we mark the pod as failed? today if we don't write the creds in the file we return an error code that exec can see. That way the broker knows when to stop looking. If the pod completes, but no secret exists does this mean there are no secrets or that we failed to write them?

Reply:
Yes. The APB would fail if we couldn't create the secret.